File: test/ConfigTest.java
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import java.util.HashMap;
import java.util.Map;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ConfigTest {

  private static Map<String, String> base() {
    var env = new HashMap<String, String>();
    env.put("EVENTS_URL", "https://example.invalid/events.json");
    env.put("MAIL_TO", "liste@jug-da.de");
    env.put("MAIL_FROM", "orga@jug-da.de");
    env.put("SMTP_HOST", "smtp.example.invalid");
    env.put("SMTP_USER", "user");
    env.put("SMTP_PASS", "pass");
    return env;
  }

  private static Map<String, String> with(String key, String value) {
    var env = base();
    env.put(key, value);
    return env;
  }

  @Test
  @DisplayName("reports every missing variable at once, not just the first")
  void reportsAllMissing() {
    var thrown = assertThrows(IllegalStateException.class,
        () -> Config.load(Map.of("SMTP_HOST", "smtp.example.invalid")));
    assertTrue(thrown.getMessage().contains("EVENTS_URL"), thrown.getMessage());
    assertTrue(thrown.getMessage().contains("MAIL_TO"), thrown.getMessage());
    assertTrue(thrown.getMessage().contains("MAIL_FROM"), thrown.getMessage());
  }

  @Test
  @DisplayName("defaults to submission on 587 with STARTTLS")
  void defaultsToSubmission() {
    var smtp = Config.load(base()).mail().smtp();
    assertEquals(587, smtp.port());
    assertFalse(smtp.secure());
    assertTrue(smtp.authenticates());
  }

  @Test
  @DisplayName("treats 465 as implicit TLS")
  void implicitTlsOn465() {
    assertTrue(Config.load(with("SMTP_PORT", "465")).mail().smtp().secure());
  }

  @Test
  @DisplayName("SMTP_SECURE overrides the port convention in both directions")
  void secureOverride() {
    var forced = base();
    forced.put("SMTP_PORT", "587");
    forced.put("SMTP_SECURE", "true");
    assertTrue(Config.load(forced).mail().smtp().secure());

    var disabled = base();
    disabled.put("SMTP_PORT", "465");
    disabled.put("SMTP_SECURE", "false");
    assertFalse(Config.load(disabled).mail().smtp().secure());
  }

  @Test
  @DisplayName("rejects a nonsense port rather than failing later at connect time")
  void rejectsBadPort() {
    assertThrows(IllegalStateException.class, () -> Config.load(with("SMTP_PORT", "submission")));
    assertThrows(IllegalStateException.class, () -> Config.load(with("SMTP_PORT", "99999")));
  }

  @Test
  @DisplayName("allows a relay that authorises by IP instead of credentials")
  void allowsUnauthenticatedRelay() {
    var env = base();
    env.remove("SMTP_USER");
    env.remove("SMTP_PASS");
    assertFalse(Config.load(env).mail().smtp().authenticates());
  }

  @Test
  @DisplayName("a username without a password is a misconfiguration, not anonymous auth")
  void halfCredentialsRejected() {
    var env = base();
    env.remove("SMTP_PASS");
    assertThrows(IllegalStateException.class, () -> Config.load(env));
  }

  @Test
  @DisplayName("Mastodon is optional but must be configured in full")
  void mastodonAllOrNothing() {
    assertNull(Config.load(base()).mastodon());
    assertThrows(IllegalStateException.class, () -> Config.load(with("MASTO_URL", "https://m.example")));
    assertThrows(IllegalStateException.class, () -> Config.load(with("MASTO_TOKEN", "t")));

    var full = base();
    full.put("MASTO_URL", "https://m.example");
    full.put("MASTO_TOKEN", "t");
    assertEquals("https://m.example", Config.load(full).mastodon().url());
  }

  @Test
  @DisplayName("DRY_RUN accepts the usual spellings and defaults off")
  void dryRunSpellings() {
    assertFalse(Config.load(base()).dryRun());
    for (var yes : new String[] {"true", "TRUE", "1", "yes", "on"}) {
      assertTrue(Config.load(with("DRY_RUN", yes)).dryRun(), yes);
    }
    for (var no : new String[] {"false", "0", "no", ""}) {
      assertFalse(Config.load(with("DRY_RUN", no)).dryRun(), no);
    }
  }

  @Test
  @DisplayName("an unset Actions variable arrives as an empty string, not as absent")
  void blankCountsAsAbsent() {
    var env = base();
    env.put("SMTP_PORT", "");
    env.put("SMTP_SECURE", "");
    env.put("MASTO_URL", "");
    env.put("MASTO_TOKEN", "");
    var config = Config.load(env);
    assertEquals(587, config.mail().smtp().port());
    assertFalse(config.mail().smtp().secure());
    assertNull(config.mastodon());
  }

  @Test
  @DisplayName("the state file defaults into the checkout and can be redirected")
  void stateFileConfigurable() {
    assertEquals("state/sent.json", Config.load(base()).stateFile().toString());
    assertEquals("/tmp/x.json", Config.load(with("STATE_FILE", "/tmp/x.json")).stateFile().toString());
  }
}

File: test/config.test.js
import assert from 'node:assert/strict';
import { test } from 'node:test';
import * as config from '../src/config.js';

const base = {
  EVENTS_URL: 'https://example.invalid/events.json',
  MAIL_TO: 'liste@jug-da.de',
  MAIL_FROM: 'orga@jug-da.de',
  SMTP_HOST: 'smtp.example.invalid',
  SMTP_USER: 'user',
  SMTP_PASS: 'pass',
};

test('reports every missing variable at once, not just the first', () => {
  assert.throws(
    () => config.load({ SMTP_HOST: 'smtp.example.invalid' }),
    /EVENTS_URL, MAIL_TO, MAIL_FROM/,
  );
});

test('defaults to submission on 587 with STARTTLS', () => {
  const { smtp } = config.load(base).mail;
  assert.equal(smtp.port, 587);
  assert.equal(smtp.secure, false);
  assert.deepEqual(smtp.auth, { user: 'user', pass: 'pass' });
});

test('treats 465 as implicit TLS', () => {
  assert.equal(config.load({ ...base, SMTP_PORT: '465' }).mail.smtp.secure, true);
});

test('SMTP_SECURE overrides the port convention in both directions', () => {
  assert.equal(config.load({ ...base, SMTP_PORT: '587', SMTP_SECURE: 'true' }).mail.smtp.secure, true);
  assert.equal(config.load({ ...base, SMTP_PORT: '465', SMTP_SECURE: 'false' }).mail.smtp.secure, false);
});

test('rejects a nonsense port instead of silently using NaN', () => {
  assert.throws(() => config.load({ ...base, SMTP_PORT: 'submission' }), /SMTP_PORT must be a port number/);
});

test('allows a relay that authorises by IP instead of credentials', () => {
  const { SMTP_USER, SMTP_PASS, ...noAuth } = base;
  assert.equal(config.load(noAuth).mail.smtp.auth, undefined);
});

test('a username without a password is a misconfiguration, not anonymous auth', () => {
  const { SMTP_PASS, ...partial } = base;
  assert.throws(() => config.load(partial), /SMTP_PASS/);
});

test('Mastodon is optional but must be configured in full', () => {
  assert.equal(config.load(base).mastodon, null);
  assert.throws(() => config.load({ ...base, MASTO_URL: 'https://m.example' }), /MASTO_TOKEN/);
  assert.throws(() => config.load({ ...base, MASTO_TOKEN: 't' }), /MASTO_URL/);
  assert.deepEqual(config.load({ ...base, MASTO_URL: 'https://m.example', MASTO_TOKEN: 't' }).mastodon,
    { url: 'https://m.example', accessToken: 't' });
});

test('DRY_RUN accepts the usual truthy spellings and defaults off', () => {
  assert.equal(config.load(base).dryRun, false);
  for (const value of ['true', 'TRUE', '1', 'yes', 'on']) {
    assert.equal(config.load({ ...base, DRY_RUN: value }).dryRun, true, `expected ${value} to enable dry run`);
  }
  for (const value of ['false', '0', 'no', '']) {
    assert.equal(config.load({ ...base, DRY_RUN: value }).dryRun, false, `expected ${value} to disable dry run`);
  }
});

test('an unset Actions variable arrives as an empty string, not undefined', () => {
  // GitHub renders ${{ vars.MISSING }} as '', which must fall back to the defaults.
  const cfg = config.load({ ...base, SMTP_PORT: '', SMTP_SECURE: '', DRY_RUN: '', MASTO_URL: '', MASTO_TOKEN: '' });
  assert.equal(cfg.mail.smtp.port, 587);
  assert.equal(cfg.mail.smtp.secure, false);
  assert.equal(cfg.dryRun, false);
  assert.equal(cfg.mastodon, null);
});
